### PR TITLE
Handle "form_post" OAuth and OpenID responses (fix for Apple OpenID)

### PR DESCRIPTION
**Pull request by aidenfoxx**
## Description

This should allow both standard and "form_post" response modes from OpenID and OAuth providers. Adding the following option to a provider config will trigger the new flow: `AUTH_PROVIDER_PARAMS="{"response_mode":"form_post"}"`.

This is required by Apple to return the user email.

Also removed a superfluous check in the router which is covered by `getUserID`, and add a missing "state" variable check related to that.

## Type of Change

- [ ] Bugfix
- [x] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [ ] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [x] Documentation was added/updated. PR: https://github.com/directus/docs/pull/198/files

### CODE REVIEW

File: api/src/auth/drivers/oauth2.ts [hunk 1]
```diff
@@ -254,6 +255,15 @@ export function createOAuth2AuthRouter(providerName: string): Router {
...
 
+	router.post(
+		'/callback',
+		bodyParser.urlencoded({ extended: false }),
+		(req, res) => {
+			res.redirect(303, `./callback?${new URLSearchParams(req.body)}`);
+		},
+		respond
+	);
+
```
Comment (aidenfoxx) [inline api/src/auth/drivers/oauth2.ts:266]:
```suggestion

```

File: api/src/auth/drivers/openid.ts [hunk 2]
```diff
@@ -278,6 +279,15 @@ export function createOpenIDAuthRouter(providerName: string): Router {
...
 
+	router.post(
+		'/callback',
+		bodyParser.urlencoded({ extended: false }),
+		asyncHandler(async (req, res, next) => {
+			res.redirect(303, `./callback?${new URLSearchParams(req.body)}`);
+		}),
+		respond
+	);
+
```
Comment (aidenfoxx) [inline api/src/auth/drivers/openid.ts:290]:
```suggestion
	router.post(
		'/callback',
		bodyParser.urlencoded({ extended: false }),
		(req, res) => {
			res.redirect(303, `./callback?${new URLSearchParams(req.body)}`);
		},
		respond
	);

```

**Comment by aidenfoxx** [comment]
@rijkvanzanten I've tested this with "form_post" via Google, but I've not tested Apple (because fuck paying £79 for the privilege). SMH Apple! 😠 

If you're able to test the configuration in the attached Docs PR, it would be appreciated.

File: api/src/auth/drivers/oauth2.ts [hunk 3]
```diff
@@ -254,6 +255,14 @@ export function createOAuth2AuthRouter(providerName: string): Router {
 		respond
 	);
 
+	router.post(
+		'/callback',
+		bodyParser.urlencoded({ extended: false }),
+		(req, res) => {
+			res.redirect(303, `./callback?${new URLSearchParams(req.body)}`);
```
Comment (aidenfoxx) [inline api/src/auth/drivers/oauth2.ts:262]: 303 ensures the redirect will be GET. https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**